datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -1161,6 +1164,15 @@ impl SMJStream {
             let filter_columns = if chunk.buffered_batch_idx.is_some() {
                 if matches!(self.join_type, JoinType::Right) {
                     get_filter_column(&self.filter, &buffered_columns, &streamed_columns)
+                } else if matches!(self.join_type, JoinType::LeftSemi) {
+                    let buffered_columns = self.buffered_data.batches
+                        [chunk.buffered_batch_idx.unwrap()]
+                    .batch
+                    .columns()
+                    .iter()
+                    .map(|column| take(column, &buffered_indices, None))
+                    .collect::<Result<Vec<_>, ArrowError>>()?;
```

Comment:
This is a bit duplicate to above one block where preparing `buffered_columns`, we can create a small function for it.